WD-26081/auto reject old specs

josecabralf opened the pull request:
[WD-26081](https://warthogs.atlassian.net/browse/WD-26081)

- created the auto reject script
- added Scope as a drive client config (as script needed read/write perms)

[WD-26081]: https://warthogs.atlassian.net/browse/WD-26081?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

QA Steps:
You may use [this doc](https://docs.google.com/document/d/14LUSFQPaOSvopGFpAv9-50diHhIKZwaj_OSg78ZKWOs) to test the script.

1. Remove all synced docs from the db (so as not to reject production specs accidentally)
2. Add a row in the DB with the linked document (doc id: 14LUSFQPaOSvopGFpAv9-50diHhIKZwaj_OSg78ZKWOs)
3. Run the taskfile scripts

josecabralf committed:
Added initial version of script

josecabralf committed:
cleanup

josecabralf committed:
Added Scope as a drive config

josecabralf committed:
More cleanup
- Removed unnecessary migration
- Removed unnecessary change on permissions when dryRun

josecabralf committed:
Reject script runs once a day

The following changes were also done:
- CleanupID and Timestamp removed from global reject config, and instead define during run to prevent inconsistencies.
-	Reordered reject/main.go main() method to more closely ressemble the sync script.
- Split long lines into multiple ones.
- Changed visibility of FindStaleSpecs to private.

josecabralf committed:
Use go routines for reject script

josecabralf committed:
Update stale docs update date

josecabralf committed:
Added reject service
- Polished reject/main.go code

josecabralf committed:
Merge branch 'main' into auto-reject-old-specs

josecabralf committed:
update charmcraft

josecabralf committed:
Fixed updater

josecabralf committed:
Addressed PR comments

josecabralf committed:
Created optional param file-id for easier testing.

josecabralf committed:
Addressed useful copilot comments

alnvdl-work commented on cmd/reject/main.go:
A nitpick for other future projects: it would be a good idea to have a single `specs` command with 4 sub-commands (`specs api`, `specs migrate`, `specs sync`, and now `specs reject`). This make it less likely that we will end up with conflicting commands when one runs `go install ./...` in their machines.

This also makes it easier to document what each sub-command is doing. For a proper CLI experience, we could either use something like https://github.com/juju/cmd (widely used at Canonical) or https://github.com/spf13/cobra (widely used in the industry), but that's to be discussed in the team.

josecabralf replied:
This is my first task in the project, so I didn't have the chance to set this up. Maybe we can refactor a bit in the future (?)

It sounds quite interesting... do you have any examples of project with this set up to check out?

alnvdl-work replied:
Many :)

Using juju/cmd: https://github.com/canonical/ua-contracts/blob/develop/cmd/contract/cmd.go

Using cobra: https://github.com/canonical/contracts/blob/main/cmd/cctl/cctl/cctl.go

Using go-flags: https://github.com/canonical/pebble/blob/master/internals/cli/cli.go

Most importantly than the tools, there's a new design guideline for writing CLI commands as well: https://docs.google.com/document/d/1qsNbj1aipo6F6RuEqd3un43ArBoMenCiddCcrJBl23s/edit?tab=t.0

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,214 @@
+		DB:           db,
+		Config:       config,
+	}
+}
```
No need for this function, in Go it's perfectly fine to initialize a struct in the place it's meant to be used without a constructor. You only need a new function if you are actually performing some special verification or custom behavior.

Alternatively... you may want to have a function like this to make sure people don't forget to pass all the requirements (it's easier to forget with a struct). In that case, you should make the fields in the struct unexported.

josecabralf replied:
Interesting... Quite different from other languages!
I'll modify the umsizi project to follow this tip

alnvdl-work commented on specs/updater.go:
```diff
@@ -0,0 +1,311 @@
+//   - 1 byte for the cell start character (typically '\v')
+//
+// This is used when advancing from one cell's position to the next cell's start index.
+const CellBoundaryOffset int64 = 2
```
This const is only used inside the package, no need to export it.

alnvdl-work commented on specs/updater.go:
```diff
@@ -0,0 +1,311 @@
+const CellBoundaryOffset int64 = 2
+
+// CellCoordinates represents the position of a status cell in a Google Doc table
+type CellCoordinates struct {
```
Same here, internal type.

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,214 @@
+	var specs []*db.Spec
+	err := r.DB.
+		Where("LOWER(status) IN ?", []string{"drafting", "braindump"}).
+		Where("google_doc_updated_at < ?", time.Now().AddDate(0, -6, 0)).
```
Idea: maybe make this timeout configurable like you did for `RejectInterval`?

In the command, you could read this as a parameter like `--threshold 6m`, parse it with `ParseDuration` and then you could just pass that as a config via `RejectConfig` and use it here as a configurable parameter:
```go
...
Where("google_doc_updated_at < ?", time.Now().Add(r.config.Threshold)).
...
```

josecabralf replied:
That's a really nice point.
Should this be a command param or should it come from env vars? 
Being that this is intended to run indefinitely as a background task, I'm thinking env var better adapts to the intended use (?)

alnvdl-work replied:
Since this is a CLI tool that you control, I'd say a command line argument.

Environment variables are typically used when:
1. Deploying services as 12-factor apps for example.
2. When you don't control the CLI (e.g., let's say you have a library that can be a part of CLI applications, and that library needs to receive input regardless of the application supporting it via arguments).

---

A side note on the use of config... In most other project I worked on, it's more common to have a `config` package that you stop using as soon as possible. In other words, let's say you have a CLI tool that starts a server: that server invokes the config parsing, and then initializes everything (structs, functions, libraries) using the config values.

And that's it, nobody else in the code bases imports  that config package. Whatever they need, they ask for it explicitly as function arguments or options structs.

That prevents your config from being used everywhere in the project and from becoming a god-object. Another benefit is that it makes testing easier: you don't need to initialize the config just to unit test independent pieces of functionality.

alnvdl-work replied:
As discussed in our 1-1, this could likely be a const in the code for now,  and a CLI argument later if needed.

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+	"google.golang.org/api/drive/v3"
+)
+
+func main() {
```
For now, before we have a solution  for building CLIs, just add a comment here explaining what this command does:
```go
// main implements a command that runs as a daemon to auto-reject specs based 
// on how long they have been open. In addition to the daemon mode, it also 
// supports rejecting an individual spec, or performing a dry-run on all specs.
```

Another upside of this is that, in addition to humans, LLMs thrive when this kind of small documentation bits are available to help them make sense of a project :)

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+	flag.StringVar(&googleDocID, "google-doc-id", "", "reject a single doc (optional)")
+	flag.Parse()
+
+	c := config.MustLoadConfig()
```
Let's standardize on `cfg` for variable names for configs here and elsewhere. `c` is a bit too vague.

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+	googleDrive, err := google.NewGoogleDrive(google.Config{
+		ClientID:          "112404606310881291739",
+		ClientEmail:       "specs-reader@roadmap-270011.iam.gserviceaccount.com",
+		ClientX509CertURL: "https://www.googleapis.com/robot/v1/metadata/x509/specs-reader%40roadmap-270011.iam.gserviceaccount.com",
```
Why aren't these read from config?

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+		PrivateKey:        c.GooglePrivateKey,
+		PrivateKeyID:      c.GooglePrivateKeyID,
+		ProjectID:         "roadmap-270011",
+		Scopes:            []string{drive.DriveScope},
```
Should these also be config?

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+		specs.RejectConfig{
+			DryRun: dryRun,
+		},
+	)
```
On a side note: I'd recommend moving this and all related initialization code to a separate function:
```
func initRejectService(cfg *config.Config) *specs.RejectService {
    ...
}
```

This should help keep the main function focused on just what it does without the initialization bits. That helps humans and LLMs make better sense of the code :)

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,214 @@
+	Config       RejectConfig
+
+	FailedCount   int
+	RejectedCount int
```
No need to export these fields, they are only read here.

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,214 @@
+		"synced_at": time.Now(),
+	}
+	if err := r.DB.Model(spec).Where("id = ?", spec.ID).Updates(updateData).Error; err != nil {
+		return fmt.Errorf("failed to update spec status in database: %w", err)
```
When returning an error from an exported function, you need to decide: do I want my callers to be able to see the details or type of a specific error returned by one of my dependencies? (in this case, gorm).

The answer is usually no: you don't want the users of `RejectService` to be able to know you are using gorm (that's an implementation detail). If you do decide to return gorm errors, you should usually document them in the Godoc, for example: `May return errors of type gorm.ErrRecordNotFound`.

That is because once you make something visible to your callers, it's very hard to hide it in the future.

So in most cases, you should usually opt for `%v` instead of `%w`.

I explain that in more detail here: https://alnvdl.github.io/2023/09/07/error-handling-in-go.html

josecabralf replied:
This is quite useful... specially since I'm still getting used to the whole "errors as values" for error handling. Thanks for the tip!

alnvdl-work commented on specs/updater.go:
```diff
@@ -0,0 +1,311 @@
+	return nil
+}
+
+// addFallbackRejectionNotice creates a fallback rejection message with red text when changelog table is not available
```
Nitpick: if you look at the Go source, you will see godoc comments are usually limited to a certain number of columns IIRC. They are not super consistent, but there's a not-very-wide-limit. Example: https://cs.opensource.google/go/go/+/refs/tags/go1.25.4:src/time/format.go

In most other projects I've participated, we set a limit of 80-columns for comments in general, as it helps with readability and terminal usage. Code itself doesn't need to be limited to 80 columns.

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,214 @@
+	for _, spec := range specs {
+		select {
+		case <-ctx.Done():
+			r.Logger.Info("rejection job cancelled")
```
Technically, the job is in the `main` function of the `reject` command (that's where you logged `starting rejection job` for example).

This  is just a function the job calls. You likely want to log this line there instead. And in fact, I just checked, and you do just that :smile:

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+			logger.Info("rejection job stopped")
+			return
+		case <-ticker.C:
+			if err := rejectService.RejectAllStaleSpecs(ctx); err != nil {
```
You might want to log that another iteration is about to start? Maybe even do it in the `RejectAllStaleSpecs` function itself? For example:
```go
logger.Info("rejecting stale specs", [threshold], [dry-run])
```

alnvdl-work reviewed (commented):
An initial review with lots of comments, but this is in going a very good direction :-)

nathanclairmonte reviewed (approved):
LGTM! 🚀 Tested with doc ID `14LUSFQPaOSvopGFpAv9-50diHhIKZwaj_OSg78ZKWOs` and works both in dry run mode and non-dry run mode. Pending Allan's comments ofc

josecabralf committed:
Updated RejectService setup.
1. Removed the NewRejectService constructor
2. Made failed and rejected count private

josecabralf committed:
Reworked godocs and errors
- godoc comments < 80 cols
- errors using %v instead of %w

josecabralf committed:
Modified access for cellBoundaryOffset and cellCoordinates

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	// Set up ticker for periodic runs
+	ticker := time.NewTicker(c.GetRejectInterval())
```
As stated elsewhere: if you control the CLI, it's more common to pass these things via explicit arguments, unless you are going to run it as a kind-of 12-factor daemon kind of thing.

That's a question I have actually: how/where/when will this code run :smile:?

alnvdl-work commented on specs/reject.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+type RejectConfig struct {
+	DryRun bool // If true, will log what would be done without making changes
```
The more common format in Go would be to have it in the line before as:
```go
// DryRun makes the Reject fuctions just log instead of acting.
DryRun bool
```

No need for the "If true" bit, since in this case it should be obvious from context.

josecabralf committed:
Updated reject command.
- Added godoc comment for main method.
- Renamed config c -> cfg for clarity and standardization.
- Made google client and project id data a configuration
- Removed ClientX509CertURL from the google client as it was not being used.
- Created initRejectService to simplify main code (separation of concerns)
- Added iteration start log

josecabralf committed:
Address rock build issue where resulting image did not have the reject script.
- Removed unneeded log from reject.go

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,121 @@
+	rejectService := initRejectService(cfg, logger)
+	rejectService.Config = specs.RejectConfig{
+		DryRun: dryRun,
+	}
```
This should be moved to `initRejectService` (pass `dryRun` to it). Modifying a struct like this after initialization by another function is usually a bad idea.

alnvdl-work commented on cmd/sync/main.go:
```diff
@@ -32,12 +32,12 @@ func main() {
+		PrivateKey:   c.GooglePrivateKey,
+		PrivateKeyID: c.GooglePrivateKeyID,
+		ProjectID:    c.GoogleProjectID,
+		Scopes:       c.GetSyncGoogleDriveScopes(),
```
Was the removal of `ClientX509CertURL` intentional?

josecabralf replied:
As explained in our call, it was discovered that the ClientX509CertURL was not used anywhere in the code

alnvdl-work commented on specs/updater.go:
```diff
@@ -0,0 +1,317 @@
+				},
+			},
+		},
+	}).Context(ctx).Do()
```
At Canonical, we usually use more compact representations for these things (see [an internal doc written by members of the OCTO team](https://docs.google.com/document/d/1IbFXyeXYlfQ5GUEEScGS7pP335Cei-5cFBdAoR973pQ/edit?tab=t.0#heading=h.puywyqz78uu1) and also [Google's style guide](https://google.github.io/styleguide/go/decisions#cuddled-braces). For example, something like:
```go
[]*docs.Request{{
	DeleteContentRange: &docs.DeleteContentRangeRequest{
		Range: &docs.Range{StartIndex: cellStartIndex, EndIndex: cellEndIndex},
	},
}, {
	InsertText: &docs.InsertTextRequest{
		Location: &docs.Location{Index: cellStartIndex},
		Text:     newStatus,
	},
}}
```

This also passes `go fmt`, it's just a more compact representation.

This is one of the annoying things about Go BTW: it does have a formatter, but it's not 100% opinionated...

alnvdl-work commented on cmd/reject/main.go:
```diff
@@ -0,0 +1,121 @@
+// initRejectService initializes the Reject service with all its dependencies
+func initRejectService(cfg *config.Config, logger *slog.Logger) *specs.RejectService {
+	dbConn, err := db.NewDB(logger, cfg)
+	if err != nil {
+		logger.Error("failed to connect to database", "error", err.Error())
+		os.Exit(1)
+	}
```
Nitpick: instead of calling logger and exit here, you could just return an error:
```go
return fmt.Errorf("failed to connect to database: %v", err)
```

Then your `main` function becomes the only point in which your program can exit. This makes it easier to reason about, as you only have one function that can exit your program:
```go
rejectService, err := initRejectService(cfg, logger)
if err != nil {
    logger.Error("cannot initialize reject service", "error", err)
	os.Exit(1)
}
```

alnvdl-work commented on specs/updater.go:
```diff
@@ -0,0 +1,317 @@
+				Fields: "foregroundColor,bold",
+			},
+		},
+	}
```
Same comment here about making things more compact.

alnvdl-work reviewed (approved):
I spent a lot of time reviewing the code itself, but I didn't run it locally (no credentials yet). But LGTM from the little I know about the project yet. Other reviewer's reviews should likely carry much more weight :)

josecabralf committed:
Configuration updates
- Made threshold a cfg
- Added default values to already known vars to prevent having to further configure charm deployment

josecabralf committed:
Compacted requests and moved config initialization into initRejectService

josecabralf committed:
Edited comments for RejectConfig
